### DISCUSSION TITLE
Apparent Sparkle and mainBundle crash after downloading an update (2.2.2)

### DISCUSSION THREAD

**Discussion by jrose-eng** [sparkle-project/Sparkle#2258] [Q&A] [answered]
Sparkle developers,

I have recently been working to migrate code on a project to use Sparkle 2.2.2 (up from 1.27.1).

Background:
1) Our application is a JVM app (which is the one built with the Sparkle framework), along with a separately-bundled helper app that is in Electron. They run together, and our .dmg update packages include both the JVM and Electron apps.

2) On Sparkle 1.27.1, after downloading an update file .dmg, the Sparkle UI would prompt us to "Install and Relaunch"; after clicking that button, both the JVM and Electron apps would be quit, the .dmg containing the updated versions of each app would be extracted, and the update would work fine. We interfaced between the JVM app and Sparkle using a `.dylib` wrapper and JNI calls. We used a basic `SUUpdater` initialised to update the `mainBundle`, and a `SUUpdaterDelegate`, with a simple implementation of the method `- (NSString *)feedURLStringForUpdater:(SUUpdater *)updater` to allow us to provide the `feedURL` dynamically from our JVM app.

3) On Sparkle 2.2.2, I have switched from using the now-deprecated `SUUpdater` to use an `SPUUpdater`, `SPUStandardUserDriver`, and `SPUUpdaterDelegate`. The Updater and UserDriver are initialised with `mainBundle` for host (and for `applicationBundle` for the Updater case). The delegate performs the same `feedURL` function as before.

Scenario:
1) All works well i.e. downloading the appcast, checking the versions, UI popups when new versions are available **until** the point where the .dmg download is complete. When the download progress bar in the Sparkle UI reaches the end, the JVM app (and Sparkle?) quits very abruptly. The Electron app does **not** get closed. Unlike before, I don't get a UI prompt to "Install and Relaunch" when the download is done. For debugging, I also implemented `- (void)updater:(SPUUpdater *)updater didDownloadUpdate:(SUAppcastItem *)item` and `- (void)updater:(SPUUpdater *)updater failedToDownloadUpdate:(SUAppcastItem *)item error:(NSError *)error` in the delegate class to produce logging, but neither one of these methods is ever called (I see no logs on my machine), leading me to believe that something has gone wrong with the end-of-download flow.

From my description here, is there anything that you think could be problematic from my setup which could be causing this strange behaviour?

Many thanks for your help.

**Comment by zorgiepoo** [comment, marked as answer]
Please check Console.app if you can find any crash logs or relevant console logs to your app's process, or `Autoupdate` or `Updater.app`. Hopefully JVM doesn't obfuscate crashes.. Although if you aren't even getting download did succeed/fail callbacks, then `Autoupdate` or `Updater.app` shouldn't have spawned / be relevant yet.

Anything Sparkle related must be called on the main thread, one thing I would see that might be easy to get wrong from a JVM application.

I assume you'll run into the same issue if you use legacy `SUUpdater` in Sparkle 2? It may help you rule out incorrect API usage since experimentally other code wouldn't have to change for trying to debug this. If the app is available somewhere I might be able to give a try.

Is your app update serving a pkg or a regular app bundle inside the dmg? I don't understand that much how the Electron app is related here. Sparkle doesn't regularly support updating "multiple apps" at once unless one of those apps is bundled inside the other (nor regularly terminates _other_ apps).

**Reply by jrose-eng** [reply]
Hello, @zorgiepoo - thank you for the swift response.

I have just tried the update flow again, and interestingly, there is some output in the logs in Console.app that suggests a `SIGABRT` is occurring right as the download is completed.

```
> Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGABRT)
Exception Codes:       KERN_PROTECTION_FAILURE at 0x0000000104901008
Exception Codes:       0x0000000000000002, 0x0000000104901008
Exception Note:        EXC_CORPSE_NOTIFY

VM Region Info: 0x104901008 is in 0x104901000-0x104902000;  bytes after start: 8  bytes before end: 4087
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      shared memory               1048fd000-104901000    [   16K] r--/r-- SM=SHM  
--->  VM_ALLOCATE                 104901000-104902000    [    4K] ---/rwx SM=NUL  
      VM_ALLOCATE                 104902000-104903000    [    4K] r--/rwx SM=PRV  

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	    0x7ff81483000e __pthread_kill + 10
1   libsystem_pthread.dylib       	    0x7ff8148661ff pthread_kill + 263
2   libsystem_c.dylib             	    0x7ff8147b1d24 abort + 123
3   libsystem_c.dylib             	    0x7ff8147b10cb __assert_rtn + 314
4   Sparkle                       	       0x1a1c061f7 -[SUInstallerLauncher pathForBundledTool:extension:fromBundle:].cold.2 + 35 (SUInstallerLauncher.m:328)
5   Sparkle                       	       0x1a1bf8d5a -[SUInstallerLauncher pathForBundledTool:extension:fromBundle:] + 587 (SUInstallerLauncher.m:328)
6   Sparkle                       	       0x1a1bf9268 __149-[SUInstallerLauncher launchInstallerWithHostBundlePath:updaterIdentifier:authorizationPrompt:installationType:allowingDriverInteraction:completion:]_block_invoke + 363 (SUInstallerLauncher.m:424)
7   libdispatch.dylib             	    0x7ff8146ae0cc _dispatch_call_block_and_release + 12
8   libdispatch.dylib             	    0x7ff8146af317 _dispatch_client_callout + 8
9   libdispatch.dylib             	    0x7ff8146bbc78 _dispatch_main_queue_drain + 943
10  libdispatch.dylib             	    0x7ff8146bb8bb _dispatch_main_queue_callback_4CF + 31

```

This is with built Sparkle as it appears on the tag 2.2.2. I'll give it a try again with the original `SUUpdater` API we had been using in 1.27.1, but this snippet I have copied here may suggest what could be going wrong in my environment. Unfortunately, the app is not publicly available, so I don't have a way of sending it to you to test first hand.

Our .dmg update is serving up regular bundled apps (the JVM app, and within that, the Electron app bundle); unless the upgrade flow for terminating apps in Sparkle 2.x has changed from 1.x, I expect it should still work to our needs (i.e. just specifying the JVM mainBundle will be enough to terminate both apps and allow the .dmg extraction to replace both app bundles), so I don't have any worries on that part right now.

**Reply by zorgiepoo** [reply]
Sparkle is failing to find its auxiliary `Autoupdate` tool inside Sparkle's framework bundle. Can I assume you did not opt into using Sparkle's XPC Services for sandboxing? Relevant assert that is failing is here: https://github.com/sparkle-project/Sparkle/blob/2.x/InstallerLauncher/SUInstallerLauncher.m#L328 . The `bundle` passed is `[NSBundle bundleForClass:[self class]]` which should be Sparkle's own framework bundle (or its own XPC Service if you opted into using them).

You might also need to make sure you're copying the framework correctly, e.g, preserving symlinks. There should be an `Autoupdate` symlink inside of `Sparkle.framework/` pointing to `Versions/B/Autoupdate` (and similar case with `Updater.app`)

**Reply by zorgiepoo** [reply]
Also this crash/assert happens after the download so you should still be getting the download delegate callbacks.

**Reply by zorgiepoo** [reply]
> (i.e. just specifying the JVM mainBundle will be enough to terminate both apps and allow the .dmg extraction to replace both app bundles)

Sparkle 1.x [just calls `[NSApp terminate:self]`](https://github.com/sparkle-project/Sparkle/blob/master/Sparkle/SUBasicUpdateDriver.m#L702) (or waits for termination in case of automatic silent installs) inside the app it's in. It does not "terminate other apps". And it [only installs one app from the dmg archive that it extracts](https://github.com/sparkle-project/Sparkle/blob/master/Sparkle/SUInstaller.m#L32) (the closest matching one based on filename and bundle ID). You might have additional logic somewhere else..

**Reply by jrose-eng** [reply]
Correct - I have not opted into using XPC Services. Having said that, I do see this log from Autoupdate; is this relevant? `Autoupdate: (Security) [com.apple.securityd:trustsettings] TrustSettingsUseXPC is enabled (via feature flags)`

Some good news: I had a look at my bundled app and realised that there were indeed some missing symlinks; I added them in, and the downloader no longer crashes. It now progresses to the extraction step. Thanks for that!

Less positive news: At the end of that process, I now hit this error:

`Autoupdate: [org.sparkle-project.Sparkle:Sparkle] Error: Timeout: agent connection was never initiated (null) (URL (null))
`
`(Sparkle) [org.sparkle-project.Sparkle:Sparkle] Error: An error occurred while running the updater. Please try again later. The remote port connection was invalidated from the updater. For additional details, please check Console logs for Autoupdate. If your application is sandboxed, please also ensure Installer Connection & Status entitlements are correctly set up: https://sparkle-project.org/documentation/sandboxing/ (URL (null))`

Looks like I've made a silly error somewhere by not initialising something correctly, but I can't quite tell from reading the code what I'm missing. Do you have any suggestions for what would be worth investigating?

**Reply by zorgiepoo** [reply]
> Correct - I have not opted into using XPC Services. Having said that, I do see this log from Autoupdate; is this relevant? Autoupdate: (Security) [com.apple.securityd:trustsettings] TrustSettingsUseXPC is enabled (via feature flags)

No.

> Some good news: I had a look at my bundled app and realised that there were indeed some missing symlinks; I added them in, and the downloader no longer crashes. It now progresses to the extraction step. Thanks for that!

Ensure `Updater.app` is there too and Updater.app is copied correctly.

> Looks like I've made a silly error somewhere by not initialising something correctly, but I can't quite tell from reading the code what I'm missing. Do you have any suggestions for what would be worth investigating?
 Write

Looks like Updater.app didn't contact Autoupdate (reached a timeout). Check logs for Updater process or from your own application. At least you can figure out if the Updater agent process is even spawned

**Reply by jrose-eng** [reply]
> Ensure Updater.app is there too and Updater.app is copied correctly.

I have a symlink `Sparkle.framework/Updater.app` to `Sparkle.framework/Versions/B/Updater.app`, as well as `Sparkle.framework/Autoupdate` to `Sparkle.framework/Versions/B/Autoupdate`. These were the symlinks I added that fixed the original crash. I also have symlinks for `Headers`, `Resources`, and `Sparkle`, pointing to their respective namesakes in `Versions/B/`. Sadly, I have still been running into trouble with the timeouts. I _think_ those are all the correct symlinks?

> Looks like Updater.app didn't contact Autoupdate (reached a timeout). Check logs for Updater process or from your own application. At least you can figure out if the Updater agent process is even spawned

I've just tried to run the updater again, with `Console.app` open and filtering on the word `updater` in the stream logs. Unfortunately, nothing more than the above log (i.e. `An error occurred while running the updater. Please try again later. The remote port connection was invalidated from the updater...`) came up. It doesn't look like the `Updater.app` agent is ever launched?

**Reply by zorgiepoo** [reply]
LaunchServices for me logs this check-in out in `Updater` process once it starts (on macOS 12.5.1 (21G83)). There's some other system messages too.

```
default	17:44:45.689728-0700	Updater	CHECKIN: pid=24934
```

The framework / your app / dylib should log out one of these format strings if an error occurs with starting this job.

```
Failed to submit progress tool job
Submit progress error: %@
```

You might be able to cross-check with the Sparkle Test App in our binary distributions.

edit: also, if you zip up the Sparkle.framework that does get copied into your app bundle I can take a look and see if anything is wrong with that. Maybe you can `codesign --verify --deep <sparkle-path>` on it as well although I'm not sure that will flag anything if you're re-signing the framework.

**Reply by jrose-eng** [reply]
Hello, @zorgiepoo - apologies for the delayed response. Many thanks for your great assistance so far.

So, I have revisited our build processes and fixed up all the problems that existed with code signing and symlink re-creation. When we spoke earlier in the week, I had simply been adding the symlinks into our bundled app directory manually. Unfortunately, I still hit the same error as earlier after the EdDSA signature verification step.

I have since done some more digging into the `launchd` logs and spotted the following, which I think explain the root cause of the issue. This is strange, because I set `Updater.app` to have `+rx` permissions when I copied it over into our app's bundled `Sparkle.framework`, so I don't know what this is about exactly. It also looks like it's referring to the Updater in the downloaded update bundle, rather than the old app I have running on my machine and that I am trying to update? Is that expected?

Here are the errors (with device names and app names replaced with ellipses, as they aren't important):

```
<Warning>: Could not find and/or execute program specified by service: 13: Permission denied: /.../org.sparkle-project.Sparkle/Launcher/gyGrbpzTE/Updater.app/Contents/MacOS/Updater
2022-09-08 16:49:19.115322 (gui/501/... [2981]) <Error>: Service could not initialize: posix_spawn(/.../Library/Caches/.../org.sparkle-project.Sparkle/Launcher/gyGrbpzTE/Updater.app/Contents/MacOS/Updater) not accessible error: 0x6f: Invalid or missing Program/ProgramArguments
```

Also, here is the `Sparkle.framework.zip`, that I believe has the correct symlinks and signing, available here: 
[Sparkle.framework.zip](https://github.com/sparkle-project/Sparkle/files/9528371/Sparkle.framework.zip)

**Reply by zorgiepoo** [reply]
> It also looks like it's referring to the Updater in the downloaded update bundle, rather than the old app I have running on my machine and that I am trying to update? Is that expected?

I think you're just inferring this wrong. Sparkle copies Updater.app from the old app into a caches directory before spawning it which is what you're seeing. Because it's important for this app to not be removed underneath from itself when installing an update to your app and relaunching it.

Clearly the missing executable bit is the issue from the `Sparkle.framework` you uploaded:

```
~/D/P/Sparkle (2.x)> ls -l /Users/msp/Downloads/Sparkle.framework/Versions/B/Updater.app/Contents/MacOS/Updater
-rw-r--r--@ ...
```

It's supposed to be:

```
-rwxr-xr-x@ ...
```

`codesign --verify --deep` unfortunately does not verify file permissions or a missing executable bit so it won't help you catch this type of issue. For some reason your `Autoupdate` executable is fine.

So you still need to fix how you're copying the framework.

**Reply by zorgiepoo** [reply]
If you're shelling out to an external tool, `ditto` should do the right thing with copying files.

**Reply by jrose-eng** [reply]
Running `chmod +x` on `Updater` solved the problem; I just tested the updater again, and it worked seamlessly! In terms of the permissions, I think I've spotted the problem from our build scripts, so I'll go and rectify that.

Again, @zorgiepoo, thank you so much for being a wealth of information and for taking your valuable time to provide this kind of support. I've learned a lot and I now understand a lot more about how Sparkle works/what I am doing with it!

Have a great day and keep up the fantastic work with this project.
